EIPS/eip-4972.md

```diff
@@ -0,0 +1,143 @@
+    ///  multiple operators per owner.
+    function setApprovalForAllFromName(bytes memory _owner, address _operator, bool _approved) public returns(bool);
+}
+```
+
+#### IERC20NOA
+```
+interface IERC20NOA is IERC20, INameOwnedAccount {
+    /// @notice Transfers _value amount of tokens from name  _from to address _to, 
+    /// @dev Throws unless `msg.sender` is the owner of _owner. Throw if _owner is not
+    ///  the current owner. Throw if the balance of _from does not have enough tokens to 
+    ///  spend. Emits the Transfer event.  
+    function transferFromName(bytes memory _from, address _to, uint256 _value) public returns(bool);
+
+    /// @notice Allows _spender to withdraw from _owner multiple times, up to
+    ///  the _value amount.
+    ///  @dev Throws unless `msg.sender` is the owner of _owner. Throw if _owner is 
+    ///  not the current owner. If this function is called again it overwrites the current
+    ///  allowance with _value.
+    function approveFromName(bytes memory _owner, address _spender, uint256 _value) public returns(bool);
+}
+```
+
+### Authentication
+The transfer and approve function is authenticated if and only if the message sender is the owner of the name.
+
+## Rationale
+We use bytes array to represent a name to ensure it’s flexible enough to deal with different use cases. E.g. one can encode the name service contract address the name belongs to into the bytes array. One can also encode extra authentication data, such as zero knowledge proofs, into the bytes array. In the future, we may propose a standard to formalize the name for wider adoption.
+
+The isNameOwner function is sufficient for authenticating the message sender. One can verify the owner by looking up the name owner from a name service, or check zero knowledge proofs encoded in name to prove the ownership directly without looking up anything.
+
+The addressOfName interface decouples the implementation from specific hashing algorithms, as long as the generated address doesn’t collide with EOA/CA address space.
+
+## Backwards Compatibility
+The new account type is compatible with existing ERC token standards.
+
+## Reference Implementation
+### Name Format
+The decoded format of bytes name is not defined at this standard. One straightforward implementation would be:
+```
+bytes memory name = abi.encode((string, ‘address’), (username, nameService))
+```
+where the username is the string representation of the username and nameService is the name service contract address. This will decouple the implementation from specific name services such as ENS.
+
+### Name Derived Address (INameOwnedAccount.addressOfName())
+With the bytes format mentioned above, we can follow the same rule of CREATE2 opcode to compute the NOA address from nameService and hash of the username. By restricting the deployer address as address zero we can ensure it won’t clash with existing smart contract account addresses:
```

Comment:
"By restricting the deployer address as address zero" => "By replacing the deployer address to hash of “eip-4972.addressOfName”"

Reply:
This part has been reworded.